packages/backend-deployer/src/cdk_deployer.ts

```diff
@@ -80,7 +81,23 @@ export class CDKDeployer implements BackendDeployer {
     const synthTimeSeconds = Math.floor((Date.now() - startTime) / 10) / 100;
 
     // then run type checks
-    await this.invokeTsc(deployProps);
+    try {
+      await this.invokeTsc(deployProps);
+    } catch (typeError: unknown) {
+      if (
+        synthError &&
+        typeError instanceof AmplifyError &&
+        typeError.cause?.message.match(
+          /Cannot find module '\$amplify\/env\/.*' or its corresponding type declarations/
+        )
+      ) {
+        // synth has failed and we don't have auto generated function environment definition files. This
+        // resulted in the exception caught here, which is not very useful for the customers.
+        // We instead throw the synth error for customers to fix what caused the synth to fail.
+        throw synthError;
+      }
+      throw typeError;
```

Comment:
I do wonder if we should just always throw synthError if that happens and refrain from running TSC.

Or perhaps, if both yielded errors maybe we should throw https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AggregateError and somewhat display both messages together.

Reply:
The reason we run TS is to produce better errors for customers as I mentioned in the overview. I thought about aggregating it, but in most cases when it's just duplicate with different wording which gets very confusing.